File: src/wall-height.test.js
import test from "node:test";
import assert from "node:assert/strict";
import {initial,step} from "./game.js";
import {velocityUnit} from "./attributes.js";
const unit=22.4/576;
test("high and low balls use original wall insets on all four walls",()=>{
 for(const axis of ["x","z"])for(const sign of [-1,1])for(const high of [false,true])for(const beyond of [false,true]){
  const s=initial();for(const p of s.players)p.stun=100;
  const center=axis==="x"?576:320;
  const boundary=(center-(high?24:32))*unit;
  Object.assign(s.ball,{owner:-1,lastTouch:-1,x:0,z:4,h:high?1.75:.75,flightKind:2,flightStage:high?3:1,flightIndex:high?8:0,flightFraction:0,vx:0,vz:0,speedTimer:100,nextSlowdown:0});
  s.ball[axis]=sign*(center-(beyond?20:28))*unit;
  s.ball["v"+axis]=sign*8*velocityUnit;
  const before=s.ball[axis];step(s,.04,{},[true,true]);
  const reflected=!high||beyond;
  assert.equal(Math.sign(s.ball["v"+axis]),reflected?-sign:sign,`${axis} high=${high} beyond=${beyond}`);
  const expected=(reflected?sign*boundary:before)+s.ball["v"+axis]*.04;
  assert.ok(Math.abs(s.ball[axis]-expected)<1e-8);
 }
});

File: server/wall_height_test.go
package main

import (
	"math"
	"testing"
)

func TestWallHeightInsets(t *testing.T) {
	for _, axis := range []string{"x", "z"} {
		for _, sign := range []float64{-1, 1} {
			for _, high := range []bool{false, true} {
				for _, beyond := range []bool{false, true} {
					s := initial()
					for i := range s.Players {
						s.Players[i].Stun = 100
					}
					center := 320.
					if axis == "x" {
						center = 576
					}
					inset := 32.
					stage, index := 1, 0
					if high {
						inset = 24
						stage, index = 3, 8
					}
					position := (center - 28) * terrainUnit
					if beyond {
						position = (center - 20) * terrainUnit
					}
					s.Ball = Ball{Owner: -1, LastTouch: -1, Z: 4, H: .25 + float64(stage)*.5, FlightKind: 2, FlightStage: stage, FlightIndex: index, SpeedTimer: 100}
					if axis == "x" {
						s.Ball.X = sign * position
						s.Ball.VX = sign * 8 * velocityUnit
					} else {
						s.Ball.Z = sign * position
						s.Ball.VZ = sign * 8 * velocityUnit
					}
					s.simulate(.04, [2]Input{}, [2]bool{true, true})
					value, velocity := s.Ball.Z, s.Ball.VZ
					if axis == "x" {
						value, velocity = s.Ball.X, s.Ball.VX
					}
					reflected := !high || beyond
					expected := sign * position
					expectedSign := sign
					if reflected {
						expected = sign * (center - inset) * terrainUnit
						expectedSign = -sign
					}
					if math.Copysign(1, velocity) != expectedSign || math.Abs(value-(expected+velocity*.04)) > 1e-8 {
						t.Fatal(axis, sign, high, beyond, value, velocity)
					}
				}
			}
		}
	}
}
